File: pages/base_page.py
class BasePage:
    def __init__(self, driver):
        self.driver = driver

    def find_element(self, locator):
        return self.driver.find_element(*locator)

    def find_elements(self, locator):
        return self.driver.find_elements(*locator)

    def click(self, locator):
        self.find_element(locator).click()

    def enter_text(self, locator, text):
        self.find_element(locator).clear()
        self.find_element(locator).send_keys(text)

    def get_text(self, locator):
        return self.find_element(locator).text

    def is_element_present(self, locator):
        try:
            self.find_element(locator)
            return True
        except:
            return False

File: pages/basePage.js
// BasePage class for Gypsy Nurse application
class BasePage {
  constructor(page) {
    this.page = page;
  }

  // Base actions that can be used by all page objects
 async findElement(locator) {
    return this.page.locator(locator);
  }

  async scrollToElement(locator) {
    try {
      const element = this.page.locator(locator);
      await element.scrollIntoViewIfNeeded();
    } catch (error) {
      console.log(`Could not scroll to element: ${locator}`, error.message);
      // Continue execution even if scrolling fails
    }
  }

  async click(locator) {
    // Scroll to element before clicking
    await this.scrollToElement(locator);
    await this.page.locator(locator).click();
 }

  async enterText(locator, text) {
    // Scroll to element before entering text
    await this.scrollToElement(locator);
    const element = this.page.locator(locator);
    
    await element.clear();
    await element.fill(text);
  }

  async getText(locator) {
    // Scroll to element before getting text
    await this.scrollToElement(locator);
    return await this.page.locator(locator).textContent();
  }

  async isElementPresent(locator) {
    try {
      // Scroll to element before checking visibility
      await this.scrollToElement(locator);
      const element = this.page.locator(locator);
      return await element.isVisible();
    } catch (error) {
      return false;
    }
  }
}

module.exports = BasePage;
